pds4indextools/pds4_create_xml_index.py

```diff
@@ -222,40 +217,190 @@ def process_schema_location(file_path):
+                                            if x.num is not None)))
+        renumber_map = {x: f'<{i+1}>' for i, x in enumerate(unique_nums)}
+        renumber_map[None] = ''
+
+        # We further group these by unique parent (including the number)
+        # and recursively process all children for each unique parent.
+        # When the child map is returned, we update our map using the number
+        # remapping for the current parent combined with the child map.
+        for parent, parent_group in groupby(prefix_group_list,
+                                            lambda x: x.parent):
+            parent_group_list = list(parent_group)
+
+            # Find all the entries that have children, package them up,
+            # and call renumber_xpaths recursively to renumber the next level
+            # down.
+            children = [x for x in parent_group_list if x.child is not None]
+            if children:
+                child_map = renumber_xpaths([x.child for x in children])
+                xpath_map.update(
+                    {f'{x.parent}/{x.child}':
+                        f'{x.prefix}{renumber_map[x.num]}/{child_map[x.child]}'
+                            for x in children}
+                )
+
+            # Find all the entries that have no children. These are leaf
+            # nodes. Renumber them.
+            no_children = [x for x in parent_group_list if x.child is None]
+            xpath_map.update(
+                    {f'{x.parent}': f'{x.prefix}{renumber_map[x.num]}'
+                        for x in no_children}
+            )
+
+    return xpath_map
+
+
+def split_into_elements(xpath):
+    """Extract elements from an XPath in the order they appear.
 
     Inputs:
-        xml_results    A dictionary containing XML data to be processed.
-        key            The key representing the XML tag to be processed.
-        root           The root element of the XML tree.
-        namespaces     A dictionary containing XML namespace mappings.
-        prefixes       A dictionary containing XML namespace prefixes.
-        args           Command-line arguments.
+        xpath    The XPath of a scraped element
+
+    Returns:
+        The tuple of elements the XPath is composed of.
     """
-    if args.xpaths:
-        key_new = convert_header_to_xpath(root, key, namespaces)
-        for namespace in prefixes.keys():
-            if namespace in key_new:
-                key_new = key_new.replace(
-                    '{'+namespace+'}', prefixes[namespace]+':')
-        xml_results[key_new] = xml_results[key]
-        del xml_results[key]
-    else:
-        key_new = convert_header_to_tag(key, root, namespaces)
-        for namespace in prefixes.keys():
-            if namespace in key_new:
-                key_new = key_new.replace(
-                    '{'+namespace+'}', prefixes[namespace]+':')
-        xml_results[key_new] = xml_results[key]
-        del xml_results[key]
+    elements = ()
```

Comment:
This is a real nit, but since tuples are immutable they are not meant to be appended to. If you append an object to a tuple with `+=`, it creates a whole new tuple from scratch, copying over all the old data. However, lists are mutable and are meant to be extended. So `x.append(y)` is an efficient operation, while `x += (y,)` (where x is a tuple) is less efficient.